dra-evolution/pkg/api/capacity_types.go

```diff
@@ -69,28 +74,94 @@ type ResourcePoolSpec struct {
 	// them) empty pool.
 }
 
+const ResourcePoolMaxSharedCapacity = 128
 const ResourcePoolMaxDevices = 128
 
+type DeviceShape struct {
+	// Attributes defines the attributes of this device shape.
+	// The name of each attribute must be unique.
+	//
+	// Must not have more than 32 entries.
+	//
+	// +listType=atomic
+	// +optional
+	Attributes []DeviceAttribute `json:"attributes,omitempty" protobuf:"bytes,3,opt,name=attributes"`
```

Comment:
So these become shared attributes that all partitions share? No ability to define e.g. the common set of attributes *just* for e.g. `1g.5gb` devices that exist at different memory slices.

Reply:
yeah, those still get layered in on a per-partition basis

Reply:
Sure, but I want to be able to define e.g.
```
deviceAttributeGroups:
  - name: common-attributes
    items: ...
  - name: common-mig-1g.5gb-attributes
    items: ...
  - name: common-mig-2g.10gb-attributes
    items: ...
```
And then layer them in as appropriate.

Reply:
In other words, partitions individually have attributes. So, in the "device" that is presented to the claim side of the model, you have `shape.Attributes + partition.Attributes + device.Attributes` available to CEL.

I do NOT go so far as to define a "PartitionShape" which common-izes attributes from a particular type of Partition. That is possible but there are diminishing returns. Option 5 does effectively do that (and more).

Reply:
This is a byproduct of nesting. If everything is top-level and referenceable by a name then it "just works" and is fully flexible.

Reply:
Maybe can you create an Option 6 with what you are thinking here?

Reply:
I really don't see what we get by nesting like this. At the end of the day this structure is something generated by a driver for the scheduler / auto-scaler to consume (there is no human in the loop). So in my opinion we should be optimizing for (1) size, (2) flexibility, (3) ease of filling out, (4) ease of consuming.

Having top level `DeviceAttributeGroups` as well as top level `SharedCapacityGroups` (which is what I proposed in my [very first prototype](https://github.com/kubernetes-sigs/wg-device-management/blob/main/nv-partitionable-resources/pkg/resource/new/api.go)) meets all of these goals. It reduces size significantly, makes things fully flexible, is easy for me as the vendor to fill out, and is easy for the scheduler / autoscaler to consume.

The only "weirdness" comes from these things being top-level, but I don't actually see that being too much of a mental hurdle for driver implementors (especially with good documentation and complete examples).

Reply:
The goal of the "shape" captures the shared capacity / shared capacity consumed model as well as the common attributes *once* and re-use it across all devices. That's where the big size win comes in, not just with attributes but with the partition model. I am not clear on how you plan to capture both of those, it would be useful to see.

Reply:
I tried top-level shared capacity groups in Option 2, it actually made things LARGER. Maybe I did it wrong? I don't think I did. I think because they are groups per-GPU, we list them over and over. Whereas with the nesting, we don't need to list them, because they are implicit in the structure.

Maybe we can do that implicit duplication with shared groups at the top level? I am not sure how we know which to do that with and which to as independent. Maybe a flag? We also need to capture the repeated "consumption" models, rather than repeat them on a per-device basis, to get the full effect.

I think that's much harder to understand, frankly, and the only thing I see us gaining is a theoretically ability to have an "independent" shared group that is across the slice not on the per-device basis. Which is something we could always add later at the slice level, if we needed it. We have no need for it at this point, that I know of.

Reply:
This is somewhat off topic, but something you said here made me think of this....

I haven't discussed it much, but MIG actually supports a second level of nesting. I think it could still be modeled with this Option 4 (the second level of nesting still pulls from the same pool of resources).

However, we want to limit selection of devices from this second level of nesting to a single claim (because they all share access to the same physical memory). Meaning if someone requests e.g. a `'1c.3g.20gb' device in two different claims, they MUST come from different parent MIGs.

i.e.
```
apiVersion: resource.k8s.io/v1alpha2
kind: ResourceClaim
metadata:
  name: mig-devices-and-nics
spec:
  requests:
  - name: mig-0
    deviceClassName: gpu.nvidia.com
    deviceSelectors: "device['type'] == 'MIG' && device['profile'] == '1c.3g.20gb'"
    
  - name: mig-1
    deviceClassName: gpu.nvidia.com
    deviceSelectors: "device['type'] == 'MIG' && device['profile'] == '1c.3g.20gb'"

  - name: mig-2
    deviceClassName: gpu.nvidia.com
    deviceSelectors: "device['type'] == 'MIG' && device['profile'] == '1c.3g.20gb'"
    
  requirements:
  - matchAttribute: parentMIG
```

We currently support this in our "classic" DRA driver (because we control the API through our CRD), but I don't see an easy way to support this restriction with structured parameters.

Reply:
Any thougts on this?

Reply:
So, do these nested partitions really need to be visible in K8s? Are they offered to different containers? Could this be part of the config?

Since they are not consumable across claims, I don't see how K8s even knowing about them matters; the scheduler wouldn't have any flexibility but to give a while MIG anyway. The only flexibility would be association with different containers. But maybe the driver can handle that for these (don't associate the MIG with any specific container).